# Abstract #

---


A tutorial introduction to computational modeling of Wnt signaling pathway in a human colorectal cancer dataset using static Bayesian network models is provided. This work endeavours to expound in detail the simulation study in Matlab along with the code while explaining the concepts related to Bayesian networks. This is done in order to ease the understanding of beginner students and researchers in transition to computational signaling biology who intend to work in the field of modeling of signaling pathways. The case study is based on the contents of the advance article by Sinha(2014) and takes the reader in a step by step process of how the collection and the transformation of the available biological information from literature is done, the integration of the heterogeneous data and prior biological knowledge in the network, the simulation study is designed, the hypothesis regarding a biological phenomena transformed into computational framework, and results and inferences drawn using d-connectivity/separability are reported. It is hoped that the walkthrough will aid biologists understand the design of the computational experiments using causal models. The manuscript finally ends with a programming assignment to help the readers get hands on experience of a perturbation project.

# Files and Downloads #

---


  * The following can be found in [[File-link](https://drive.google.com/folderview?id=0B7Kkv8wlhPU-T05wTTNodWNydjA&usp=sharing)]

  1. .m files (contains function definitions)
  1. .mat files (contains data)
  1. unpublished manuscript [[link](http://biorxiv.org/content/early/2014/11/04/011064)]

Please use the citations provided in the link below, if this work or a part of it is used for further research and publication purpose.

  * Description regarding the files can be found in [[Downloads-link](https://code.google.com/p/static-bn-for-wnt-signaling-pathway/wiki/Downloads)]

# Acknowledgements #

---


  1. Thanks to **The Royal Society of Chemistry (RSC)** for giving permission to reproduce parts of material in Shriprakash Sinha, Integr. Biol., 2014, DOI: 10.1039/C4IB00124A.
  1. This work is a part of the **Netherlands Bioinformatics Centre** (NBIC) BioRange II project BR2.5 (code - IGD71G). Netherlands Bioinformatics Centre was supported by the **Netherlands Genomics Initiative** (NGI) via the Netherlands Genomics Institute in The Netherlands.
  1. **Dr. ir. R. H. J. Fastenau** (Dean of Faculty of Electrical Engineering Mathematics and Computer Science), **Dr. Prof. Ir. K. Ch. A. M. Luyben** (Rector Magnificus) and **Drs. D. J. van den Berg** (President) at Delft University of Technology, for providing support for conducting this work and giving permission to submit the manuscript.
  1. **Dr. Wim Verhaegh** (faculty at Netherlands Bioinformatics Centre and a principal scientist at Molecular Diagnostic Lab in Philips Research) for providing technical details of Naive Bayes model for replicating experiments in Verhaegh et al. 2.
  1. **Dr. Marcel J. T. Reinders** (scientific director of Bioinformatics Research at Netherlands Bioinformatics Centre and a professor at Delft University of Technology) for refining the proposed model and suggesting inclusion of methylation data.
  1. **Dr. Robert P.W. Duin** (retired associate professor at Delft University of Technology) for informal discussions on 2-holdout experiments
  1. **Dr. Jeroen de Ridder** (assistant professor at Delft University of Technology) and Phd candidate **Marc Hulsman** (Delft University of Technology), for informal discussions on estimation of probability values on methylation data.
  1. **Mr. Prabhat Sinha** and **Mrs. Rita Sinha** for financially supporting this project while the author was on educational and work leave.